authoring-service/src/main/kotlin/com/devotion/authoring/service/DefaultSurveyAuthoringService.kt

```diff
@@ -30,44 +31,63 @@ class DefaultSurveyAuthoringService(private val surveyRepository: SurveyReposito
     private val answersDtoListType = object : TypeToken<List<AnswerIdAndText>>() {}.type
 
     override fun addQuestion(@NotEmpty surveyId: String, @Valid question: QuestionText) {
-        kafkaTemplate.sendGenericMessage(ModifyQuestionEvent(Action.CREATE, surveyId, null, question), kafkaConfig.questionCapturedTopic)
+        kafkaTemplate.sendGenericMessage(AddQuestionEvent(surveyId, question), kafkaConfig.questionCapturedTopic)
     }
 
     override fun updateQuestion(@NotEmpty surveyId: String, @NotNull questionId: String, @Valid question: QuestionText) {
-        kafkaTemplate.sendGenericMessage(ModifyQuestionEvent(Action.UPDATE, surveyId, questionId, question), kafkaConfig.questionCapturedTopic)
+        kafkaTemplate.sendGenericMessage(UpdateQuestionEvent(surveyId, questionId, question), kafkaConfig.questionCapturedTopic)
     }
 
     override fun deleteQuestion(@NotEmpty surveyId: String, @NotNull questionId: String) {
-        kafkaTemplate.sendGenericMessage(ModifyQuestionEvent(Action.DELETE, surveyId, questionId), kafkaConfig.questionCapturedTopic)
+        kafkaTemplate.sendGenericMessage(DeleteQuestionEvent(surveyId, questionId), kafkaConfig.questionCapturedTopic)
     }
 
+
     // todo: make sure to not retry on validationexception!!!
     @KafkaListener(topics = ["\${kafka.questionCapturedTopic}"], containerFactory = "jsonKafkaListenerContainerFactory", errorHandler = "validationErrorHandler")
-    fun onModifyQuestionEvent(event: ModifyQuestionEvent) {
+    fun onModifyQuestionEvent(event: QuestionEvent) {
         val survey = getValidSurvey(event.surveyId)
         validateModifyEvent(event, survey)
-        val newQuestion = modelMapper.map(event.question, Question::class.java)
-        if (event.questionId != null)
-            newQuestion.id = event.questionId
-        when (event.action) {
-            Action.CREATE -> survey.questions.add(newQuestion)
-            Action.UPDATE -> survey.questions[survey.questions.indexOfFirst { it.id == newQuestion.id }] = newQuestion
-            Action.DELETE -> survey.questions.removeAt(survey.questions.indexOfFirst { it.id == newQuestion.id })
+        when (event) {
```

Comment:
We have a couple of places where we have `when` block, which violates **open-closed** principle. I'm wondering if we can use **polymorphism** instead?

Reply:
👍  any suggestions in shape of code?

Reply:
What does *shape of code* mean? :) Code looks good in general